README.md
MirrorDispatcher implements [Dispatcher] by introspecting a class instance
using [dart:mirrors]. dart:mirrors allows you to invoke an instance's methods
by their string names.

Construct a [MirrorDispatcher] with a class instance.
MirrorDispatcher.dispatch("someMethod") will return a Future of whatever
value it returns.

This is for JSON-RPC, so there are some restrictions. A method can only be invoked with 
positional or named parameters, not both. Optional parameters should work as expected. 

MirrorDispatcher is mainly a wrapper around dart:mirrors.
For any method dispatched, Dispatcher returns either the return value of
the method or instances of one of three "Exception" classes. Most errors
will be corralled into these objects, so that runtime exceptions don't get
thrown, instead are returned in an orderly manner.


## Usage

A simple usage example:

```dart
import 'package:mirror_dispatcher/mirror_dispatcher.dart';

/// Look in test/mirror_dispatcher_test.dart for additional examples.

/// Make a class with the API we want. Something silly but instructive.
///
/// Hold count in Foo.
class Foo {
  num _count = 0;

  /// initialize with a count
  Foo(this._count);

  /// increment with an optional positional parameter
  void increment([num aNumber = 1]) => _count += aNumber;

  /// decrement with a required named parameter
  void decrement({required num aNumber}) => _count -= aNumber;

  /// get the current value of count
  num getCount() => _count;

  /// we decrement by 2 a lot.
  num goTwoLess() {
    _count -= 2;
    return _count;
  }
}

void main() async {
  /// make a dispatcher with an initialized instance of Foo class
  var dispatcher = MirrorDispatcher(Foo(29));

  /// dispatch a method with a parameter
  await dispatcher.dispatch('increment', 4);

  /// get the new value
  var c = await (dispatcher.dispatch('getCount'));
  printCount(c); // 33

  /// dispatch a method with a named parameter
  await dispatcher.dispatch('decrement', {'aNumber': 18});
  var d = await (dispatcher.dispatch('getCount'));
  printCount(d); // 15

  /// dispatch a method without a parameter
  await dispatcher.dispatch('increment');
  var e = await (dispatcher.dispatch('getCount'));
  printCount(e); // 16

  var f = await dispatcher.dispatch('goTwoLess');
  printCount(f); // 14
}

void printCount(num aCount) {
  print('the currentCount is $aCount');
}

```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/jwashin/mirror_dispatcher


CHANGELOG.md
## 1.0.0

- Initial version, created by Stagehand

## 1.0.1

- pub.dev wanted changes for publication


lib/src/mirror_dispatcher_base.dart
import 'dart:async';
import 'dart:mirrors';

import 'package:rpc_dispatcher/rpc_dispatcher.dart';
import 'package:rpc_exceptions/rpc_exceptions.dart';

/// MirrorDispatcher implements [Dispatcher] by introspecting a class instance
/// using [dart:mirrors]. dart:mirrors allows you to invoke an instance's methods
/// by their string names.
///
/// Construct a [MirrorDispatcher] with a class instance.
/// MirrorDispatcher.dispatch("someMethod") will return a Future of whatever
/// value it returns.
/// It's mainly a wrapper around reflect.
/// For any method dispatched, Dispatcher returns either the return value of
/// the method or instances of one of three "Exception" classes. Most errors
/// will be corralled into these objects, so that runtime exceptions don't get
/// thrown, instead are returned in an orderly manner.
class MirrorDispatcher implements Dispatcher {
  /// the initialized class instance we will be performing methods on.
  dynamic instance;

  /// constructor
  MirrorDispatcher(this.instance);

  @override
  Future<dynamic> dispatch(String methodName,
      [dynamic positionalParams, Map<String, dynamic>? namedParams]) async {
    var posParams = positionalParams ?? [];
    if (posParams is Map<String, dynamic>) {
      namedParams = positionalParams;
      posParams = [];
    }
    if (posParams is! List) {
      posParams = [posParams];
    }
    namedParams = namedParams ?? <String, dynamic>{};
    var symbolMap = <Symbol, dynamic>{};
    if (namedParams.isNotEmpty) {
      symbolMap = symbolizeKeys(namedParams);
    }

    var instanceMirror = reflect(instance);

    Symbol methodMirror;
    try {
      methodMirror = getMethodMirror(instanceMirror, methodName);
    } on MethodNotFoundException catch (_) {
      return MethodNotFoundException('Method not found: $methodName');
    }

    InstanceMirror t;
    try {
      t = instanceMirror.invoke(methodMirror, posParams, symbolMap);
    } on TypeError catch (e) {
      return InvalidParametersException('$e');
    } on NoSuchMethodError catch (e) {
      return InvalidParametersException('$e');
    } on RuntimeException catch (e) {
      return e;
    } catch (e) // Errors might crash the server
    {
      return RuntimeException('$e');
    }
    var resp = await t.reflectee;
    return resp;
  }
}

/// Find and return the method in the class of the mirror of the instance.
/// Throw MethodNotFound if the methodName is private or an attribute
/// or not found,
Symbol getMethodMirror(dynamic instanceMirror, String methodName) {
  ClassMirror classMirror = instanceMirror.type;
  for (var classMember in classMirror.declarations.keys) {
    var instanceMethod = MirrorSystem.getName(classMember);
    if (instanceMethod == methodName) {
      var methodMirror = classMirror.declarations[classMember];
      if (methodMirror is! MethodMirror || methodMirror.isPrivate) {
        throw MethodNotFoundException();
      }
      return classMember;
    }
  }
  throw MethodNotFoundException();
}


example/mirror_dispatcher_example.dart
import 'package:mirror_dispatcher/mirror_dispatcher.dart';

/// Look in test/mirror_dispatcher_test.dart for additional examples.

/// Make a class with the API we want. Something silly but instructive.
///
/// Hold count in Foo.
class Foo {
  num _count = 0;

  /// initialize with a count
  Foo(this._count);

  /// increment with an optional positional parameter
  void increment([num aNumber = 1]) => _count += aNumber;

  /// decrement with a required named parameter
  void decrement({required num aNumber}) => _count -= aNumber;

  /// get the current value of count
  num getCount() => _count;

  /// we decrement by 2 a lot.
  num goTwoLess() {
    _count -= 2;
    return _count;
  }
}

void main() async {
  /// make a dispatcher with an initialized instance of Foo class
  var dispatcher = MirrorDispatcher(Foo(29));

  /// dispatch a method with a parameter
  await dispatcher.dispatch('increment', 4);

  /// get the new value
  var c = await (dispatcher.dispatch('getCount'));
  printCount(c); // 33

  /// dispatch a method with a named parameter
  await dispatcher.dispatch('decrement', {'aNumber': 18});
  var d = await (dispatcher.dispatch('getCount'));
  printCount(d); // 15

  /// dispatch a method without a parameter
  await dispatcher.dispatch('increment');
  var e = await (dispatcher.dispatch('getCount'));
  printCount(e); // 16

  var f = await dispatcher.dispatch('goTwoLess');
  printCount(f); // 14
}

void printCount(num aCount) {
  print('the currentCount is $aCount');
}
